File: 프로그래머스/lv2/86971. 전력망을 둘로 나누기/전력망을 둘로 나누기.py
from collections import deque

def solution(n, wires):
    answer = 1e9
            
    tree = [[] for _ in range(n+1)]
    for a, b in wires:
        tree[a].append(b)
        tree[b].append(a)
    
    
    # 시작점, 끊을 노드
    def bfs(start, cut):
        cnt = 1
        q = deque()
        q.append(start)
        
        visited = [False] * (n+1)
        visited[start] = True
        
        while q:
            now = q.popleft()
            for next in tree[now]:
                # now에 연결되어있는 노드 중 방문했거나 끊은 노드면 패스 
                if next == cut or visited[next]:
                    continue
                visited[next] = True
                q.append(next)
                cnt += 1
        
        return cnt
    
    
    for i, j in wires:
        cnt = bfs(i, j)
        answer = min(answer, abs(cnt - (n-cnt)))
    
    return answer

File: 프로그래머스/2/86971. 전력망을 둘로 나누기/전력망을 둘로 나누기.java
import java.util.*;

class Solution {
    static ArrayList<Integer>[] graph;
    static int answer = 100;
    
    public int solution(int n, int[][] wires) {
        graph = new ArrayList[n+1];
        for(int i=0; i<n+1; i++) {
            graph[i] = new ArrayList<>();
        }
        
        for(int[] w : wires) {
            int v1 = w[0];
            int v2 = w[1];
            graph[v1].add(v2);
            graph[v2].add(v1);
        }
        
        for(int[] w : wires) {
            int i = w[0];
            int j = w[1];
            
            // wires 순회하면서 연결 하나씩 끊어서 bfs 돌려보기
            int cnt = bfs(n, i, j);
            answer = Math.min(answer, Math.abs(cnt - (n-cnt)));
        }
        
        return answer;
    }
    
    // n, 시작 지점, 자를 전선 번호 
    public int bfs(int n, int start, int cut) {        
        Queue<Integer> q = new LinkedList<>();
        boolean[] visited = new boolean[n+1];
        q.offer(start);
        visited[start] = true;
        int cnt = 1;
        
        while(!q.isEmpty()) {
            int now = q.poll();
            for(int next : graph[now]) {
                // 이미 방문한 곳이거나 자를 곳이면 continue 
                if(visited[next] == true || next == cut) continue;
                q.offer(next);
                visited[next] = true;
                cnt++;
            }
        }
        return cnt;
    }
}